Add Dataset.from_spark

maddiedawson opened the pull request:
Adds static method Dataset.from_spark to create datasets from Spark DataFrames.

This approach alleviates users of the need to materialize their dataframe---a common use case is that the user loads their dataset into a dataframe, uses Spark to apply some transformation to some of the columns, and then wants to train on the dataset.

https://github.com/huggingface/datasets/issues/5678

lu-wang-dl commented on src/datasets/packaged_modules/spark/spark.py:
```diff
@@ -0,0 +1,177 @@
+            if os.path.isfile(probe[0]):
+                return
+
+        raise ValueError("When using Dataset.from_spark on a multi-node cluster, cache_dir must specify an NFS")
```
```suggestion
        raise ValueError("When using Dataset.from_spark, all workers should be able to access cache_dir")
```

maddiedawson replied:
Done

maddiedawson commented on src/datasets/arrow_dataset.py:
```diff
@@ -1211,6 +1212,35 @@ def from_text(
+        return SparkDatasetReader(
+            df,
+            cache_dir=cache_dir,
+            redownload=redownload,
```
Done

maddiedawson commented on src/datasets/arrow_dataset.py:
```diff
@@ -1211,6 +1212,35 @@ def from_text(
+    def from_spark(
+        df: pyspark.sql.DataFrame,
+        cache_dir: str = None,
+        redownload: bool = False,
```
Done

maddiedawson commented:
@mariosasko Would you or another HF datasets maintainer be able to review this, please?

albertvillanova reviewed (commented):
Thanks for the work done, @maddiedawson.

Please note there was a previous PR adding this functionality:
- #5401

However the addition of this conversion was rejected by @lhoestq: see comment https://github.com/huggingface/datasets/pull/5401#issuecomment-1372277934
> Thinking more about it I don't really see how those two methods help in practice, since one can already do datasets <-> pandas <-> spark and those two methods don't add value over this.

lhoestq commented:
Amazing ! Great job @maddiedawson 

Do you know if it's possible to also support writing to Parquet using the HF ParquetWriter if `file_format="parquet"` ?

Parquet is often used when people want to stream the data to train models - which is suitable for big datasets. On the other hand Arrow is generally used for local memory mapping with random access.

> Please note there was a previous PR adding this functionality

Am I right to say that it uses the spark workers to prepare the Arrow files ? If so this should make the data preparation fast and won't fill up the executor's memory as in the previously proposed PR

maddiedawson commented:
Thanks for taking a look! Unlike the previous PR's approach, this implementation takes advantage of Spark mapping to distribute file writing over multiple tasks. (Also it doesn't load the entire dataset into memory :) )

Supporting Parquet here sgtm; I'll modify the PR.

I also updated the PR description with a common Spark-HF use case that we want to improve.

maddiedawson commented:
Hey @albertvillanova @lhoestq , would one of you be able to re-review please? Thank you!

lhoestq commented on src/datasets/io/spark.py:
```diff
@@ -0,0 +1,40 @@
+        **kwargs,
+    ):
+        super().__init__(
+            path_or_paths="unused",
```
I guess you can modify `AbstractDatasetReader`'s typing to allow `None` here
```suggestion
            path_or_paths=None,
```

maddiedawson replied:
Done

lhoestq commented on tests/test_arrow_dataset.py:
```diff
@@ -3629,6 +3630,39 @@ def test_dataset_from_generator_features(features, data_generator, tmp_path):
+        ("3", 3, 3.0),
+    ]
+    df = spark.createDataFrame(data, "col_1: string, col_2: int, col_3: float")
+    dataset = Dataset.from_spark(df)
```
Could you add a test to make sure that calling `from_spark` twice in a row with two different DataFrames results in two different Dataset objects ?

The `DatasetBuilder` uses a `cache_dir` with subdirectories defined after the name of the builder ("spark") and the arguments of the dataset to generate. Here since the `**kwargs` passed to `DatasetBuilder.__init__` won't change between two `from_spark`, it may reuse the same cache and return the same dataset.

If this is the case we can add a `df_id` parameter to the `SparkConfig`, which can be set to to be the python `id` of the Spark DataFrame for example (or any identifier of the Spark DataFrame).

maddiedawson replied:
I modified the DatasetBuilder to cache based on the dataframe's semanticHash and added a test for it. This isn't perfect; e.g if a user generates a dataset from query "SELECT * FROM T", adds rows to the table, and then regenerates the dataset, the semantic hash for the dataframe will be the same, so the user will get the stale data. However, they can pass the force_download=True option to from_spark to force redownloading. Wdyt?

lhoestq replied:
Looks good ! (I was off the last few days, sorry for the late reply)

maddiedawson replied:
Unfortunately no, the same df will generate a different a semantic hash in different Spark sessions.

lhoestq reviewed (commented):
Good job with this, I left a few comment :)

maddiedawson commented on src/datasets/arrow_dataset.py:
```diff
@@ -57,6 +57,7 @@
 import pandas as pd
 import pyarrow as pa
 import pyarrow.compute as pc
+import pyspark
```
Done

maddiedawson commented on src/datasets/arrow_dataset.py:
```diff
@@ -1211,6 +1212,43 @@ def from_text(
+        force_download: bool = False,
+        **kwargs,
+    ):
+        """Create Dataset from Spark DataFrame.
```
Done

maddiedawson commented on src/datasets/io/spark.py:
```diff
@@ -0,0 +1,40 @@
+    def read(self):
+        download_mode = DownloadMode.FORCE_REDOWNLOAD if self._force_download else None
+        self.builder.download_and_prepare(download_mode=download_mode)
+        # TODO: Support as_streaming_dataset.
```
Done. My first follow-up task after this PR will be to add support for as_streaming_dataset

maddiedawson commented on tests/test_arrow_dataset.py:
```diff
@@ -3629,6 +3630,39 @@ def test_dataset_from_generator_features(features, data_generator, tmp_path):
+    assert isinstance(dataset, Dataset)
+    assert dataset.num_rows == 1
+    assert dataset.num_columns == 2
+    assert dataset.column_names == ["idx", "img_bytes"]
```
Done

maddiedawson commented:
@lhoestq this is ready for another pass! Thanks so much 🙏

maddiedawson commented:
Friendly ping @lhoestq , also cc @polinaeterna who may be able to help take a look?

maddiedawson commented on src/datasets/arrow_dataset.py:
```diff
@@ -1211,6 +1212,47 @@ def from_text(
+        features: Optional[Features] = None,
+        cache_dir: str = None,
+        force_download: bool = False,
+        file_format: str = "arrow",
```
Done

maddiedawson commented on src/datasets/arrow_dataset.py:
```diff
@@ -1211,6 +1212,47 @@ def from_text(
+                Whether to force download of the dataset. When not forcing download, if a dataset is present in the
+                cache, it will be reused.
+            file_format (`str`, defaults to `"arrow"`):
+                The file format to use when writing the cache. Must be either `"arrow"` or `"parquet"`.
```
Done

maddiedawson commented on src/datasets/arrow_dataset.py:
```diff
@@ -1211,6 +1212,47 @@ def from_text(
+            features=features,
+            cache_dir=cache_dir,
+            force_download=force_download,
+            file_format=file_format,
```
Done

maddiedawson commented on src/datasets/arrow_dataset.py:
```diff
@@ -1211,6 +1212,47 @@ def from_text(
+                cache, it will be reused.
+            file_format (`str`, defaults to `"arrow"`):
+                The file format to use when writing the cache. Must be either `"arrow"` or `"parquet"`.
+        """
```
Done

maddiedawson commented on src/datasets/packaged_modules/__init__.py:
```diff
@@ -32,6 +33,7 @@ def _hash_python_lines(lines: List[str]) -> str:
     "json": (json.__name__, _hash_python_lines(inspect.getsource(json).splitlines())),
     "pandas": (pandas.__name__, _hash_python_lines(inspect.getsource(pandas).splitlines())),
     "parquet": (parquet.__name__, _hash_python_lines(inspect.getsource(parquet).splitlines())),
+    "spark": (spark.__name__, _hash_python_lines(inspect.getsource(spark).splitlines())),
```
Thanks! Done

maddiedawson commented on setup.py:
```diff
@@ -137,6 +137,8 @@
     "packaging",
     "responses<0.19",
+    # For from_spark
+    "pyspark",
```
Done

lhoestq commented:
Merging `main` into this branch should fix the CI

lhoestq reviewed (approved):
Awesome ! I'd be interested in having someone else review it and then we're good to merge :)

maddiedawson commented:
Thanks @lhoestq ! Is there a way for me to trigger the github workflow myself to triage the test failure? I'm not able to repro the test failures locally.

maddiedawson commented:
There were two test issues in the workflow that I wasn't able to reproduce locally:

- Python 3.7: createDataFrame fails due to a pickling error. I modified the tests to instead write and read from json files
- Python 3.10: A worker crashes for unknown reasons. I modified the spark setup to explicitly specify local mode in case it was trying to do something else; let's see if that fixes the issue

maddiedawson commented:
Also one more question @lhoestq when is the next datasets release? We're hoping this can make it in

lhoestq commented:
I just re-ran the CI.
I think we can do a release right after this PR is merged ;)

lu-wang-dl commented on tests/test_arrow_dataset.py:
```diff
@@ -3618,6 +3620,67 @@ def test_dataset_from_generator_features(features, data_generator, tmp_path):
     _check_generator_dataset(dataset, expected_features)
 
 
+def _create_spark_dataframe(tmp_dir):
```
Could we split the test to a different class so that we can put the spark creation into `setUp` function?

maddiedawson replied:
I reverted this change

lu-wang-dl reviewed (approved):
LGTM. Left some minor comments.

maddiedawson commented:
Thanks all! @lhoestq could we re-run CI again please? I think we have to disable this feature on python 3.7 due to the pickling error. The other failure was due to https://issues.apache.org/jira/browse/SPARK-30952 so I rewrote the df processing

maddiedawson commented on src/datasets/io/spark.py:
```diff
@@ -0,0 +1,46 @@
+        load_from_cache_file: bool = True,
+        file_format: str = "arrow",
+        **kwargs,
+    ):
```
Done

maddiedawson commented on src/datasets/filesystems/__init__.py:
```diff
@@ -54,6 +55,15 @@ def is_remote_filesystem(fs: fsspec.AbstractFileSystem) -> bool:
         return False
 
 
+def rename(fs: fsspec.AbstractFileSystem, src: str, dst: str):
```
Done

maddiedawson commented:
Thanks @lhoestq , this is ready for another CI run. I pinned the pyspark version to see if that fixes the pickling issue

maddiedawson commented:
The remaining CI issues have been addressed! They were

- dill=0.3.1.1 is incompatible with cloudpickle, used by Spark. The min-dependency tests use this dill version, and those were failing. I added a skip-test annotation to skip Spark tests when using this dill version. This shouldn't be a production issue since if users are using that version of dill, they won't really be able to do anything with Spark anyway.
- One of the Spark APIs used in this feature (mapInArrow) is incompatible with Windows. I filed a Spark ticket for the team to investigate. For the tests, I added another annotation to skip Spark tests on Windows. In the next PR (adding streaming mode), we should be able to support Windows since that won't use mapInArrow.

I ran the CI on my forked branch: https://github.com/maddiedawson/datasets/pull/2 Everything passes except one instance of tests/test_metric_common.py::LocalMetricTest::test_load_metric_frugalscore; it looks like a flake.

@lhoestq granted that the CI passes here, is this ok to merge and release? We'd like to put out a blog post tomorrow to broadcast this to Spark users!

maddiedawson commented:
Thanks @lhoestq ! Could you help take a look at the error please? Seems unrelated...

FAILED tests/test_arrow_dataset.py::BaseDatasetTest::test_map_multiprocessing_on_disk - NotADirectoryError: [WinError 267] The directory name is invalid: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmptfnrdj4x\\cache-5c5687cf5629c97a_00000_of_00002.arrow'
===== 1 failed, 2152 passed, 23 skipped, 20 warnings in 461.68s (0:07:41) =====

yanzia12138 commented:
python 3.9.2
Got an error  _pickle.PicklingError use Dataset.from_spark.

Did the dataset import load data from spark dataframe using multi-node Spark cluster
df = spark.read.parquet(args.input_data).repartition(50)
ds = Dataset.from_spark(df, keep_in_memory=True,
                        cache_dir="/pnc-data/data/nuplan/t5_spark/cache_data")
ds.save_to_disk(args.output_data)

Error : 
_pickle.PicklingError: Could not serialize object: RuntimeError: It appears that you are attempting to reference SparkContext from a broadcast variable, action, or transforma
tion. SparkContext can only be used on the driver, not in code that it run on workers. For more information, see SPARK-5063.
23/06/16 21:17:20 WARN ExecutorPodsWatchSnapshotSource: Kubernetes client has been closed (this is expected if the application is shutting down.)

lhoestq commented:
Hi @yanzia12138 ! Could you open a new issue please and share the full stack trace ? This will help to know what happened exactly
